File: backend/app/schemas/prescription.py
"""
Prescription schemas for API requests and responses.
"""
from datetime import date, datetime
from typing import List, Optional
from uuid import UUID

from pydantic import BaseModel, Field

from ..models.prescription import (
    DosageUnit, DosageFrequency, PrescriptionStatus,
    MedicationCreate, MedicationResponse,
    PrescriptionItemCreate, PrescriptionItemResponse,
    PrescriptionCreate, PrescriptionResponse, PrescriptionUpdate
)


class PrescriptionListResponse(BaseModel):
    """Response schema for prescription list."""
    prescriptions: List[PrescriptionResponse]
    total: int
    page: int
    size: int
    pages: int


class PrescriptionSearchParams(BaseModel):
    """Search parameters for prescriptions."""
    patient_id: Optional[UUID] = None
    status: Optional[PrescriptionStatus] = None
    prescribed_by: Optional[str] = None
    date_from: Optional[date] = None
    date_to: Optional[date] = None
    pharmacy_name: Optional[str] = None
    
    
class MedicationSearchParams(BaseModel):
    """Search parameters for medications."""
    name: Optional[str] = None
    generic_name: Optional[str] = None
    code: Optional[str] = None
    manufacturer: Optional[str] = None


class PrescriptionFormOutput(BaseModel):
    """Prescription form for printing/export."""
    
    # Patient Information
    patient_name: str
    patient_number: str
    patient_birth_date: date
    patient_age: int
    patient_gender: str
    patient_address: str
    patient_phone: Optional[str] = None
    
    # Insurance Information
    insurance_type: Optional[str] = None
    insurance_number: Optional[str] = None
    
    # Prescription Information
    prescription_id: int
    prescribed_date: date
    valid_until: date
    prescribed_by: str
    clinic_name: Optional[str] = "医療法人オープン電カル"
    clinic_address: Optional[str] = None
    clinic_phone: Optional[str] = None
    
    # Medical Information
    diagnosis: Optional[str] = None
    clinical_info: Optional[str] = None
    
    # Medications
    medications: List[dict] = Field(..., description="List of medications with dosage info")
    
    # Additional Information
    notes: Optional[str] = None
    pharmacy_name: Optional[str] = None
    
    # Generated Information
    total_medications: int
    generated_at: datetime = Field(default_factory=datetime.now)


class DrugInteractionCheck(BaseModel):
    """Drug interaction check result."""
    medication_a: str
    medication_b: str
    interaction_level: str  # "minor", "moderate", "major", "contraindicated"
    description: str
    recommendation: Optional[str] = None


class PrescriptionValidation(BaseModel):
    """Prescription validation result."""
    prescription_id: int
    is_valid: bool
    errors: List[str] = []
    warnings: List[str] = []
    drug_interactions: List[DrugInteractionCheck] = []
    
    
class PrescriptionStats(BaseModel):
    """Prescription statistics."""
    total_prescriptions: int
    active_prescriptions: int
    completed_prescriptions: int
    cancelled_prescriptions: int
    expired_prescriptions: int
    most_prescribed_medications: List[dict]
    prescriptions_by_month: List[dict]

File: frontend/src/types/prescription.ts
export const DosageUnit = {
  MG: "mg",
  G: "g",
  ML: "ml",
  L: "l",
  TABLET: "錠",
  CAPSULE: "カプセル",
  PACKET: "包",
  BOTTLE: "本",
  TUBE: "管",
  SHEET: "シート"
} as const;

export const DosageFrequency = {
  ONCE_DAILY: "1日1回",
  TWICE_DAILY: "1日2回", 
  THREE_TIMES_DAILY: "1日3回",
  FOUR_TIMES_DAILY: "1日4回",
  AS_NEEDED: "頓服",
  BEFORE_MEALS: "食前",
  AFTER_MEALS: "食後",
  BETWEEN_MEALS: "食間",
  BEDTIME: "就寝前"
} as const;

export const PrescriptionStatus = {
  DRAFT: "下書き",
  ACTIVE: "有効",
  COMPLETED: "完了",
  CANCELLED: "取消",
  EXPIRED: "期限切れ"
} as const;

export type DosageUnitType = typeof DosageUnit[keyof typeof DosageUnit];
export type DosageFrequencyType = typeof DosageFrequency[keyof typeof DosageFrequency];
export type PrescriptionStatusType = typeof PrescriptionStatus[keyof typeof PrescriptionStatus];

export interface Medication {
  id: number;
  name: string;
  generic_name?: string;
  code?: string;
  unit: DosageUnitType;
  manufacturer?: string;
  contraindications?: string;
  side_effects?: string;
  interactions?: string;
  created_at: string;
  updated_at: string;
}

export interface PrescriptionItem {
  id: number;
  medication_id: number;
  medication: Medication;
  dosage: number;
  dosage_unit: DosageUnitType;
  frequency: DosageFrequencyType;
  duration_days: number;
  total_amount: number;
  instructions?: string;
}

export interface Prescription {
  id: number;
  patient_id: string;
  prescribed_date: string;
  valid_until: string;
  status: PrescriptionStatusType;
  clinical_info?: string;
  diagnosis?: string;
  notes?: string;
  prescribed_by: string;
  pharmacy_name?: string;
  created_at: string;
  updated_at: string;
  items: PrescriptionItem[];
}

export interface PrescriptionCreate {
  patient_id: string;
  prescribed_date: string;
  valid_until: string;
  status?: PrescriptionStatusType;
  clinical_info?: string;
  diagnosis?: string;
  notes?: string;
  prescribed_by: string;
  pharmacy_name?: string;
  items: PrescriptionItemCreate[];
}

export interface PrescriptionItemCreate {
  medication_id: number;
  dosage: number;
  dosage_unit: DosageUnitType;
  frequency: DosageFrequencyType;
  duration_days: number;
  total_amount: number;
  instructions?: string;
}

export interface DrugInteractionCheck {
  medication_a: string;
  medication_b: string;
  interaction_level: "minor" | "moderate" | "major" | "contraindicated";
  description: string;
  recommendation?: string;
}

export interface PrescriptionValidation {
  prescription_id: number;
  is_valid: boolean;
  errors: string[];
  warnings: string[];
  drug_interactions: DrugInteractionCheck[];
}

export interface PrescriptionFormOutput {
  // Patient Information
  patient_name: string;
  patient_number: string;
  patient_birth_date: string;
  patient_age: number;
  patient_gender: string;
  patient_address: string;
  patient_phone?: string;
  
  // Insurance Information
  insurance_type?: string;
  insurance_number?: string;
  
  // Prescription Information
  prescription_id: number;
  prescribed_date: string;
  valid_until: string;
  prescribed_by: string;
  clinic_name?: string;
  clinic_address?: string;
  clinic_phone?: string;
  
  // Medical Information
  diagnosis?: string;
  clinical_info?: string;
  
  // Medications
  medications: Array<{
    name: string;
    generic_name?: string;
    dosage: number;
    dosage_unit: DosageUnitType;
    frequency: DosageFrequencyType;
    duration_days: number;
    total_amount: number;
    instructions?: string;
    manufacturer?: string;
  }>;
  
  // Additional Information
  notes?: string;
  pharmacy_name?: string;
  
  // Generated Information
  total_medications: number;
  generated_at: string;
}